---
title: "oral_phyloseq"
author: "Erica Ryu"
date: "2/9/2023"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1.  Generate Phyloseq Object

The purpose of this script is to process 16S sequences through dada2 and phyloseq, and generate a phyloseq object.

## load packages
```{r}
library(ggplot2)
library(dada2)
library(phyloseq)
library(dplyr)
library(DECIPHER)
library(phangorn)
```

## set variables for loading data
```{r}
set.seed(100)
path = "data/fwd"
filt_path = "data/filt_path"
```

## filter and trim reads
```{r}
fileFwd <- list.files(path, full.names = TRUE)

ii <- sample(length(fileFwd), 3)
for(i in ii) { print(plotQualityProfile(fileFwd[i]) + ggtitle("Fwd")) }

# points to where the filtered files should go
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFwd <- file.path(filt_path, basename(fileFwd))

for(i in seq_along(fileFwd)) {
  fastqFilter(c(fileFwd[[i]]),c(filtFwd[[i]]),trimLeft=10, truncLen=c(150), maxN=0, maxEE=2, truncQ=2,compress=TRUE)
}
## Reads were truncated to remove adapters and low quality bases. All other parameters were kept default.
```

## infer sequence variants
```{r}
exists <- file.exists(filtFwd)
derepFs <- derepFastq(filtFwd[exists], verbose=TRUE)
sam.names <- sapply(strsplit(basename(fileFwd), ".fastq"), `[`, 1)
names(derepFs) <- sam.names[exists]

ddF <- dada(derepFs[1:40], err=NULL, selfConsist=TRUE)

pdf("output/ErrorRates_ddf.pdf", width=6, height=6)
plotErrors(ddF)
dev.off()

dadaFs <- dada(derepFs, err=ddF[[1]]$err_out, pool=TRUE)
```

## construct sequence table
```{r}
seqtab.all <- makeSequenceTable(dadaFs)
dim(seqtab.all) 

sum(seqtab.all) 
table(nchar(getSequences(seqtab.all)))
```

## remove chimeras
```{r}
seqtab <- removeBimeraDenovo(seqtab.all)
dim(seqtab)

sum(seqtab) 

sum(seqtab)/sum(seqtab.all)
table(nchar(getSequences(seqtab)))
```

##  track reads - bash prep
These steps are conducted in bash.
```{bash eval=FALSE, echo=T}
### get read counts for raw fastq files
# change directory to where project is located
cd Nepal_oral_microbiome

# add read counts to the file saved in the output directory
for filename in data/fwd/*.fastq
do
echo $(cat $filename|wc -l)/4|bc >> output/raw_read_counts_per_step
done

# get list of columns
echo data/fwd/*.fastq > output/raw_filename
```

## track reads - R prep
These steps are conducted in R.
```{r}
# import raw count data
raw_name <- read.table("output/raw_filename")
raw_reads <- read.table("output/raw_read_counts_per_step")

# transpose names
raw_name_col <- t(raw_name)

# clean up names
raw_name_col <- gsub("data/fwd/", "", raw_name_col)
raw_name_col <- gsub(".fastq", "", raw_name_col)

# collapse into tables
raw_count <- cbind(raw_name_col, raw_reads)

# change column names
colnames(raw_count) <- c("sample", "raw_count")

# filter out samples that did not pass initial read QC
samples_remove <- c("RAJ0019YZ_genotek", "THA0065JZ_genotek")
raw_count_mod <- raw_count[!(raw_count$sample %in% samples_remove),]
```

## track reads - R 
```{r}
# put in alphabetical order
raw_count_mod <- raw_count_mod[order(names(setNames(raw_count_mod, raw_count_mod)))]
# get read counts for all steps
getN <- function(x) sum(getUniques(x))
filtered <- as.data.frame(as.matrix(sapply(derepFs, getN)))
filtered$sample <- raw_count_mod$sample
denoised <- as.data.frame(as.matrix(sapply(dadaFs, getN)))
denoised$sample <- raw_count_mod$sample
nochim <- as.data.frame(as.matrix(rowSums(seqtab)))
nochim$sample <- raw_count_mod$sample

# merge into table
track <- merge(raw_count_mod, filtered, by = "sample", all = T)
track <- merge(track, denoised, by = "sample", all = T)
track <- merge(track, nochim, by = "sample", all = T)
colnames(track) <- c("sample", "input", "filtered", "denoised", "nochim")
head(track)
track$retained_overall <- track$nochim / track$input

# save read track table
write.csv(track, file = "output/track_reads.csv")
```

## assign taxonomy
```{r}
ref_fasta <- tempfile()
download.file("http://benjjneb.github.io/dada2/rdp_train_set_14.fa.gz", destfile = ref_fasta)
taxtab <- assignTaxonomy(seqtab, refFasta = ref_fasta)
colnames(taxtab) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")
```

## construct phylogenetic tree
```{r}
seqs <- getSequences(seqtab)
names(seqs) <- seqs

alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) 
fit = pml(treeNJ, data=phang.align)


fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
```

## construct phyloseq object
```{r}
# load in metadata
map<-read.csv("data/s1_tab_surveydata.csv")

head(map)
map<-data.frame(map)

# construct object
map <- map[!duplicated(map$SampleID),] 
all(rownames(seqtab) %in% map$SampleID) # check if TRUE

#If false, run the lines below to remove sequences without meta data and metadata without sequences
seqtab[(rownames(seqtab) %in% map$SampleID),]->seqtab
map[(map$SampleID %in% rownames(seqtab)),]->map
#this should now be TRUE
all(rownames(seqtab) %in% map$SampleID)

rownames(map) <- map$SampleID
keep.cols=names(map)
samdf <- map[rownames(seqtab), keep.cols]
ps <- phyloseq(tax_table(taxtab), 
               sample_data(samdf), 
               otu_table(seqtab, taxa_are_rows = FALSE),
               phy_tree(fitGTR$tree)
)
ps

# save phyloseq object
saveRDS(ps, "output/oralmicrobiome_ps.rds")
```

