---
title: "CCA"
author: "Erica Ryu"
date: "10/4/2022"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/ericaryu/Documents/school_stuff/grad_school/PSU/lab/Nepal/Himalayan_oral_microbiome/coding")
```

# The purpose of this script is to 

## load packages
```{r}
library(phyloseq)
library(FactoMineR)
library(factoextra)
library(dplyr)
library(vegan)
library(caret)
library(ggvegan)
library(tidyr)
library(ggplot2)
library(gridExtra)
```

## load data
```{r}
phyloseq_complete <- readRDS("ps_complete.rds")

# subset based on extraction kit
qiagen <- subset_samples(phyloseq_complete, Condition == "Qiagen")
psoil <- subset_samples(phyloseq_complete, Condition == "Psoil")

# phyloseq from cluster
# phyloseq_cluster <- readRDS("/Users/ericaryu/Downloads/oralmicrobiome_ps.rds")
# phyloseq_new <- readRDS("/Users/ericaryu/Downloads/oralmicrobiome_metadata_online_ps.rds")

phyloseq.noncontam_all <- readRDS("ps_noncontam.rds")
phyloseq <- readRDS("oralmicrobiome_metadata_online_ps.rds")

# load alpha diversity 
rfxn_micro <- read.csv("/Users/ericaryu/Documents/school_stuff/grad_school/PSU/lab/Nepal/Himalayan_oral_microbiome/coding/rarefaction_edit05152023.csv", header = TRUE)
```

## set colors
```{r}
fourcolors <- c("darkslateblue", "deepskyblue", "lightblue3", "lightsalmon")
fivecolors <- c("darkslateblue", "deepskyblue", "lightblue3", "lightsalmon" , "firebrick")
```

## set up function
```{r}
# calculate beta diversity
beta_ordinate <- function(physeq, beta_dist){
  # set up data
  ps.prop <- transform_sample_counts(physeq, function(otu) otu/sum(otu))
  ps.prop@otu_table <- na.omit(ps.prop@otu_table)
  ## calculate distance and ordinate
  if (beta_dist == "weighted unifrac"){
    ord.pcoa <- ordinate(ps.prop, method = "PCoA", distance = "unifrac", weighted = TRUE)
  } else if (beta_dist == "unweighted unifrac") {
    ord.pcoa <- ordinate(ps.prop, method = "PCoA", distance = "unifrac", weighted = FALSE)
  } else {
    ord.pcoa <- ordinate(ps.prop, method = "PCoA", distance = beta_dist)
  }
}

# set up beta diversity data for plotting
plot_beta <- function(beta, physeq){
  # extract axes
  PCOAaxes <- beta$vectors[,c(1,2,3,4)]
  # extract lifestyle column from metadata and add to vectors
  lifestyle <- physeq@sam_data$Lifestyle
  PCOAaxes_meta <- cbind(PCOAaxes, lifestyle)
  df_PCOA<- as.data.frame(as.matrix(PCOAaxes_meta))
  # change industrial to American industrial
  df_PCOA$lifestyle<- gsub("Industrial", "American Industrial", df_PCOA$lifestyle)
  # add space to Recently Settled
  df_PCOA$lifestyle <- gsub("RecentlySettled", "Recently Settled", df_PCOA$lifestyle)
  df_PCOA$lifestyle <- factor(df_PCOA$lifestyle, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial"))
  return(df_PCOA)
}
```

## set up metadata for analysis
```{r}
CA_meta <- as.data.frame(phyloseq_complete@sam_data)
CA_meta <- subset(CA_meta, Condition == "Qiagen")

df_CA <- as.data.frame(as.matrix(CA_meta))
df_CA_num <- dplyr::select(df_CA, "SEX2":"RHR")
# df_CA_num$Hypertension2 <- df_CA$Hypertension2
dim(df_CA_num)
# 69 58

### remove NAs
# filter out Americans
df_CA_num$Lifestyle <- df_CA$Lifestyle
wo_euro <- subset(df_CA_num, Lifestyle != "Industrial")

# remove columns specific to women (pregnancy, menstruation, etc)
wo_mens <- dplyr::select(wo_euro, -c("MENS2":"MISC2"))

# remove columns for bitter taste perception
wo_bitter <- dplyr::select(wo_mens, -c("BTP_CTRL2":"BTP_SBZ2"))

# remove columns for vitals
wo_vital <- dplyr::select(wo_bitter, -c("HGT2":"RHR"))

# remove geographical info
wo_geo <- dplyr::select(wo_vital, -c("LAT2":"ALT2"))

### change NAs
# change NAs in EXER_FREQ2
wo_geo$EXER_FREQ2[is.na(wo_geo$EXER_FREQ2)] <- 0

# change NAs in SICK_LOC2
wo_geo$SICK_LOC2[is.na(wo_geo$SICK_LOC2)] <- 0

# change NAs in BRUSH_FREQ2
wo_geo$BRUSH_FREQ2[is.na(wo_geo$BRUSH_FREQ2)] <- 0

# change NAs in BLKT2
blkt <- wo_geo
blkt$BLKT2[is.na(blkt$BLKT2)] <- 0

# change NA in HS
household <- blkt
household$HS2[is.na(household$HS2)] <- 2

# final df for CA
df_CA_wo_lifestyle <- dplyr::select(household, -c("Lifestyle"))
df_CA_final <- as.data.frame(sapply(df_CA_wo_lifestyle, as.numeric))
# df_CA_final <- mutate_if(df_CA_numeric, is.numeric, as.factor) # not making into a factor
dim(df_CA_final)
# 63 37
```

# make labels human readable
```{r}
colnames(df_CA_final) <- c("sex", "literacy", "location", "drinking_water", "fuel", 
                          "kitchen_loc", "toilet", "grain", "Sisnu", "fish", 
                          "meat", "black_tea", "milk_tea", "soda", "milk", 
                          "yogurt", "yogurt_freq", "fermented", "ferm_freq", "food_source", 
                          "scarcity", "smoking", "tobacco", "alcohol", "exercise", 
                          "exercise_freq", "sick_checkup", "checkup_loc", "health_travel", "meds", 
                          "ayurvedic", "brushing", "brushing_freq", "age", "education", 
                          "household", "children")
```

# CA
```{r}
rownames(df_CA_final) <- rownames(df_CA_wo_lifestyle)

set.seed(100)
ca_res <- CA(df_CA_final, ncp = 37, graph = TRUE)
print(ca_res)

# The chi square of independence between the two variables is equal to 2414.648 (p-value =  0.003765287)

# plot CA
## extract coordinates
sample_CA <- as.data.frame(as.matrix(ca_res$row$coord))
factor_CA <- as.data.frame(as.matrix(ca_res$col$coord))

## simplify to just first two dimensions
sample_CA <- sample_CA[, c("Dim 1", "Dim 2")]
factor_CA <- factor_CA[, c("Dim 1", "Dim 2")]

## rename columns for ease of use
colnames(sample_CA) <- c("Dim1", "Dim2")
colnames(factor_CA) <- c("Dim1", "Dim2")

## add lifestyle info to samples
sample_CA$lifestyle  <- ifelse(grepl("CHE",row.names(sample_CA)), "Foragers",
                              ifelse(grepl("NEW00",row.names(sample_CA)) | grepl("NEW10",row.names(sample_CA)) | grepl("THA",row.names(sample_CA)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(sample_CA)) | grepl("NEW11",row.names(sample_CA)), "Expats", "Recently Settled")))

## for better visualization
scaling_factor <- 1.07

## only keep factors that are in the top 10 
CA_labels <- factor_CA[c("Sisnu", "literacy", "fuel", "health_travel", "education", "smoking", "scarcity", "ayurvedic", "children", "grain", "alcohol", "tobacco", "sex", "location", "yogurt_freq"),]

## plot
CA_plot <- ggplot(sample_CA, aes(x = Dim1, y = Dim2)) +
  geom_point(shape = 16, size = 2, aes(color = lifestyle)) +
  scale_color_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_point(data=factor_CA, shape = 21, color = "black",
            aes(x=Dim1,
                y=Dim2))+
  geom_text(data=CA_labels, size = 3, color="black",
            aes(x=Dim1*scaling_factor, 
                y=Dim2*scaling_factor,
                label=rownames(CA_labels)))+
  labs(title = "CA",
       x = "CA1 (15.89%)",
       y = "CA2 (12.87%)")

# ggsave(file="ca_plot.svg", plot=CA_plot, width=8, height=6)

fviz_ca_biplot(ca_res)
# double check association between rows and columns
chi_square <- 2414.648
# Degree of freedom
df <- (nrow(df_CA_final) - 1) * (ncol(df_CA_final) - 1)
pval <- pchisq(chi_square, df = df, lower.tail = FALSE)

# examine eigenvalues to see what percentage of variance they explain 
eig.val <- get_eigenvalue(ca_res)
#          eigenvalue variance.percent cumulative.variance.percent
# Dim.1  0.1067863586      15.88540562                    15.88541
# Dim.2  0.0871682078      12.96703395                    28.85244
# Dim.3  0.0556647538       8.28061939                    37.13306
# Dim.4  0.0499561246       7.43141084                    44.56447
# Dim.5  0.0485230095       7.21822242                    51.78269
fviz_screeplot(ca_res, addlabels = TRUE, ylim = c(0, 20))

# what is the avg eigenvalue above which axis should be kept?
col_max <- 1/(ncol(df_CA_final)-1)
# 0.02777778
row_max <- 1/(nrow(df_CA_final)-1)
# 0.01612903

# asymmetric biplot
fviz_ca_biplot(ca_res, 
               map ="rowprincipal", arrow = c(FALSE, TRUE),
               repel = TRUE)
```

# dig into the columns/factors
```{r}
# just plotting factors
fviz_ca_col(ca_res)

# specific contributions to each axis
dim1_contrib <- fviz_contrib(ca_res, choice = "col", axes = 1, top = 10)
# Sisnu, literacy, fuel, health travel, education, smoking, scarcity, ayurvedic, children, grain
dim2_contrib <- fviz_contrib(ca_res, choice = "col", axes = 2, top = 10)
# alcohol, tobacco, smoking, scarcity, sex, location, yogurt freq, literacy, Sisnu, grain
fviz_contrib(ca_res, choice = "col", axes = 1:2, top = 10)
# sisnu, literacy, fuel, alcohol, smoking, tobacco, scarcity, health travel, education, location

# ggsave(file="dim1_contrib.svg", plot=dim1_contrib, width=8, height=5)
# ggsave(file="dim2_contrib.svg", plot=dim2_contrib, width=8, height=5)

# ggsave(file="dim1_contrib_small.svg", plot=dim1_contrib, width=4, height=2)
# ggsave(file="dim2_contrib_small.svg", plot=dim2_contrib, width=4, height=2)

```

# plot CA axes
```{r}
row_axes <- get_ca_row(ca_res)$coord
row_axes <- as.data.frame(row_axes[,c(1,2,3,4)])
# add lifestyle info
row_axes$lifestyle  <- ifelse(grepl("CHE",row.names(row_axes)), "Foragers",
                              ifelse(grepl("NEW00",row.names(row_axes)) | grepl("NEW10",row.names(row_axes)) | grepl("THA",row.names(row_axes)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(row_axes)) | grepl("NEW11",row.names(row_axes)), "Expats", "Recently Settled")))
row_axes$lifestyle <- factor(row_axes$lifestyle, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"))
# change column names
colnames(row_axes) <- c("Dim1", "Dim2", "Dim3", "Dim4", "lifestyle")

# plot
ca1_axes <- ggplot(row_axes, aes(x = lifestyle, y = as.numeric(Dim1), group = lifestyle), color = black) +
  geom_violin(alpha = 0.8, aes(fill=lifestyle)) +
  geom_boxplot(width=0.1, color="black", alpha=0.2, outlier.shape = NA) + 
  scale_y_continuous(breaks=seq(-1,1,0.2), labels = scales::comma) +
  scale_fill_manual(name=NULL,
                    values=fivecolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_jitter(size = 1, col="darkgreen", width = 0.1) +
  labs(x = "Lifestyle", y = "CA1 (15.89%)") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust=0.5)) + 
  theme(legend.position = "none")

# ggsave(file="ca1_axes.svg", plot=ca1_axes, width=4, height=3)

ca2_axes <- ggplot(row_axes, aes(x = lifestyle, y = as.numeric(Dim2), group = lifestyle), color = black) +
  geom_violin(alpha = 0.8, aes(fill=lifestyle)) +
  geom_boxplot(width=0.1, color="black", alpha=0.2, outlier.shape = NA) + 
  scale_y_continuous(breaks=seq(-1,1,0.2), labels = scales::comma) +
  scale_fill_manual(name=NULL,
                    values=fivecolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_jitter(size = 1, col="darkgreen", width = 0.1) +
  labs(x = "Lifestyle", y = "CA2 (12.87%)") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust=0.5)) + 
  theme(legend.position = "none")

# ggsave(file="ca2_axes.svg", plot=ca2_axes, width=4, height=3)
```

# are CA axes correlated with PCoA axes? are differences in lifestyle associated with differences in the microbiome?
```{r}
# prep CA data
ca.data <- as.data.frame(ca_res$row$coord)
ca.data$SampleID <- factor(row.names(ca.data))
ca.data <- ca.data[order(ca.data$SampleID),]
ca.data <- ca.data[,c(1,2,3,4)]
colnames(ca.data) <- c("CA1", "CA2", "CA3", "CA4")
dim(ca.data)
  
# prep PCoA data
pcoa_corr <- subset_samples(qiagen, Lifestyle != "Industrial")
bray_ordinate_corr <- beta_ordinate(pcoa_corr, beta_dist = "bray")
bray_plot_corr <- plot_beta(bray_ordinate_corr, pcoa_corr)
bray_plot_corr_num <-  as.data.frame(sapply(bray_plot_corr, as.numeric))
rownames(bray_plot_corr_num) <- rownames(bray_plot_corr)
colnames(bray_plot_corr_num) <- c("PCoA1", "PCoA2", "PCoA3", "PCoA4", "lifestyle")
dim(bray_plot_corr_num)

# combine
pcoa_ca <- merge(ca.data, bray_plot_corr_num, by = "row.names")
pcoa_ca$lifestyle  <- ifelse(grepl("CHE",pcoa_ca$Row.names), "Foragers",
                              ifelse(grepl("NEW00",pcoa_ca$Row.names) | grepl("NEW10",pcoa_ca$Row.names) | grepl("THA",pcoa_ca$Row.names), "Agriculturalists",
                              ifelse(grepl("NEW01",pcoa_ca$Row.names) | grepl("NEW11",pcoa_ca$Row.names), "Expats", "Recently Settled")))

### axes 1
# plot 
pcoa1_ca1 <- ggplot(pcoa_ca , aes(x = PCoA1, y = CA1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "PCoA1 vs CA1",
       x = "PCoA1",
       y = "CA1")

# ggsave(file="pcoa1_ca1_small.svg", plot=pcoa1_ca1, width=4, height=2)

# correlation
cor.test(pcoa_ca$PCoA1, pcoa_ca$CA1, alternative = "two.sided", method=c("spearman"))
#  rho = 0.05347542, p value = 0.6765


### axes 2
# plot
pcoa2_ca2 <- ggplot(pcoa_ca , aes(x = PCoA2, y = CA2*-1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "PCoA2 vs CA2",
       x = "PCoA2",
       y = "CA2")

# ggsave(file="pcoa2_ca2_small.svg", plot=pcoa2_ca2, width=4, height=2)

# correlation
cor.test(pcoa_ca$PCoA2, pcoa_ca$CA2, alternative = "two.sided", method=c("spearman"))
#  rho = 0.2748656, p-value = 0.02956


### axes 3
# plot (use rev version because otherwise neg corr)
pcoa3_ca3 <- ggplot(pcoa_ca , aes(x = PCoA3, y = CA3*-1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "PCoA3 vs CA3",
       x = "PCoA3",
       y = "CA3")

# correlation
cor.test(pcoa_ca$PCoA3, pcoa_ca$CA3, alternative = "two.sided", method=c("spearman"))
#  rho = 0.01104071, p-value = 0.9315

# ggsave(file="pcoa3_ca3_small.svg", plot=pcoa3_ca3, width=4, height=2)

### CA1 vs PCoA2]
pcoa2_ca1 <- ggplot(pcoa_ca , aes(x = PCoA2, y = CA1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "PCoA2 vs CA1",
       x = "PCoA2",
       y = "CA1")

# correlation
cor.test(pcoa_ca$PCoA2, pcoa_ca$CA1, alternative = "two.sided", method=c("spearman"))
#  rho = -0.005088326, p-value = 0.9685

# ggsave(file="pcoa2_ca1_small.svg", plot=pcoa2_ca1, width=4, height=2)

### CA2 vs PCoA1
pcoa1_ca2 <- ggplot(pcoa_ca , aes(x = PCoA1, y = CA2)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "PCoA1 vs CA2",
       x = "PCoA1",
       y = "CA2")

# correlation
cor.test(pcoa_ca$PCoA1, pcoa_ca$CA2, alternative = "two.sided", method=c("spearman"))
#  rho = -0.004416283, p-value = 0.9727

# ggsave(file="pcoa1_ca2_small.svg", plot=pcoa1_ca2, width=4, height=2)
```

# CA axes vs alpha diversity
```{r}
# subset to qiagen
qiagen_rfxn <- subset(rfxn_micro, Condition == "Qiagen")

# subset to shannon and faiths
shannon_qiagen <- subset(qiagen_rfxn, measure == "Shannon")
faiths_qiagen <- subset(qiagen_rfxn, measure == "Faiths")

# remove europeans from alpha diversity
shannon_CA <- shannon_qiagen[!grepl("EUR", shannon_qiagen$SampleID),]
faiths_CA <- faiths_qiagen[!grepl("EUR", faiths_qiagen$SampleID),]

# select alpha mean value and lifestyle
shannon_CA <- shannon_CA[, c("mean", "Lifestyle")]
rownames(shannon_CA) <- rownames(ca.data)
faiths_CA <- faiths_CA[, c("mean", "Lifestyle")]
rownames(faiths_CA) <- rownames(ca.data)

shannon_CA_comb <- merge(shannon_CA, ca.data, by = "row.names")
faiths_CA_comb <- merge(faiths_CA, ca.data, by = "row.names")

### shannon axes
# plot 
shannon_ca1 <- ggplot(shannon_CA_comb, aes(x = mean, y = CA1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = Lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "Shannon Alpha vs CA1",
       x = "Shannon Alpha",
       y = "CA1")

# ggsave(file="shannon_ca1_small.svg", plot=shannon_ca1, width=4, height=2)

shannon_ca2 <- ggplot(shannon_CA_comb, aes(x = mean, y = CA2)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = Lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "Shannon Alpha vs CA2",
       x = "Shannon Alpha",
       y = "CA2")

# ggsave(file="shannon_ca2_small.svg", plot=shannon_ca2, width=4, height=2)

### faiths axes
## plot
faiths_ca1 <- ggplot(faiths_CA_comb, aes(x = mean, y = CA1)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = Lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "Faiths Alpha vs CA1",
       x = "Faiths Alpha",
       y = "CA1")

# ggsave(file="faiths_ca1_small.svg", plot=faiths_ca1, width=4, height=2)

faiths_ca2 <- ggplot(faiths_CA_comb, aes(x = mean, y = CA2)) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = Lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  geom_smooth(method = "glm") +
  labs(title = "Faiths Alpha vs CA2",
       x = "Faiths Alpha",
       y = "CA2")

# ggsave(file="faiths_ca2_small.svg", plot=faiths_ca2, width=4, height=2)

# correlation
cor.test(faiths_CA_comb$mean, faiths_CA_comb$CA2, alternative = "two.sided", method=c("spearman"))


```


# prep OTU table for CCA
```{r}
# collapse to genus level
phy_genus <-  tax_glom(pcoa_corr, "Genus", NArm = FALSE);phy_genus

# relative abundance
phy_genus_RA <- transform_sample_counts(phy_genus, function(otu) otu/sum(otu))

df_OTU_CCA <- as.data.frame(as.matrix(phy_genus_RA@otu_table))

# log transform 
# mk_logit <- function(df_OTU_CCA) {
#   return(log10(df_OTU_CCA+0.0001))}
# log_CCA <- map_df(df_OTU_CCA, .f = mk_logit)

### add taxa names
# get taxa table
CCA_tax_table <- as.data.frame(as.matrix(phy_genus_RA@tax_table))

# change NA to unclassified taxon
CCA_tax_table$Genus <- replace_na(CCA_tax_table$Genus, "_unclassified")

# and label unclassified taxonomic level
CCA_tax_table$Class <- ifelse(is.na(CCA_tax_table$Class), CCA_tax_table$Phylum, CCA_tax_table$Class)
CCA_tax_table$Order <- ifelse(is.na(CCA_tax_table$Order), CCA_tax_table$Class, CCA_tax_table$Order)
CCA_tax_table$Family <- ifelse(is.na(CCA_tax_table$Family), CCA_tax_table$Order, CCA_tax_table$Family)

CCA_tax_table$Genus <- ifelse(CCA_tax_table$Genus == "_unclassified", paste(CCA_tax_table$Family, CCA_tax_table$Genus, sep = ""), CCA_tax_table$Genus)

# isolate to genera
CCA_genera_table <- subset(CCA_tax_table, select = c(Genus))

# add taxonomic info to OTU table
CCA_OTU_log_t <- t(df_OTU_CCA)
CCA_otu_tax <- merge(CCA_OTU_log_t, CCA_genera_table, by = 0, all = TRUE)

# remove NA_classified
CCA_otu_tax <- CCA_otu_tax[!grepl("NA_unclassified", CCA_otu_tax$Genus),]

# make Genus the rownames and remove extra columns
rownames(CCA_otu_tax) <- CCA_otu_tax$Genus
CCA_otu_tax <- subset(CCA_otu_tax, select = -c(Genus, Row.names))

# transpose back
CCA_otu_tax_t <- t(CCA_otu_tax)

df_CA_factor <- mutate_if(df_CA_final, is.numeric, as.factor)
```

# CCA
```{r}
# cca_model_all <- cca(CCA_log~., df_CA_final)
# # select variables
# final_cca_model_all <- ordistep(cca_model_all, scope = formula(cca_model_all))
# 
# # calculate VIF
# vif.cca(final_cca_model_all)

cca_model_CA2 <- cca(CCA_otu_tax_t ~ alcohol + tobacco + smoking + scarcity + sex + location + yogurt_freq + literacy + Sisnu + grain, data = df_CA_factor)

vif.cca(cca_model_CA2)

cca_model_CA2
# 23.51% of species variance explained by the lifestyle factors in the model
# CCA1    CCA2    CCA3    CCA4    CCA5    CCA6    CCA7    CCA8    CCA9   CCA10   CCA11   CCA12 
# 0.07409 0.05030 0.03879 0.03302 0.03038 0.02885 0.02573 0.02464 0.02332 0.02157 0.01828 0.01707 

# test significance of the model
anova.cca(cca_model_CA2)
#          Df ChiSquare      F Pr(>F)    
# Model    12   0.38604 1.2807  0.001 ***
# Residual 50   1.25593       
anova.cca(cca_model_CA2, by = "terms")
#             Df ChiSquare      F Pr(>F)    
# alcohol      1   0.03796 1.5113  0.004 ** 
# tobacco      1   0.02806 1.1173  0.170    
# smoking      1   0.04122 1.6409  0.001 ***
# scarcity     1   0.02916 1.1609  0.082 .  
# sex          1   0.02336 0.9301  0.676    
# location     1   0.04782 1.9039  0.001 ***
# yogurt_freq  3   0.08225 1.0915  0.114    
# literacy     1   0.02276 0.9060  0.783    
# Sisnu        1   0.03959 1.5759  0.002 ** 
# grain        1   0.03385 1.3477  0.009 ** 
anova.cca(cca_model_CA2, by = "axis")
#          Df ChiSquare      F Pr(>F)    
# CCA1      1   0.07409 2.9494  0.001 ***
# CCA2      1   0.05030 2.0025  0.019 *  
# CCA3      1   0.03879 1.5441  0.415 

### plot
plot <- autoplot(cca_model_CA2)
plot

```

# CCA final plot
```{r}
df_species  <- as.data.frame(summary(cca_model_CA2)$species[,1:2])# get the species CC1 and CC2 scores
df_sites  <- as.data.frame(summary(cca_model_CA2)$sites[,1:2])# get the sites CC1 and CC2 scores
df_environ  <- as.data.frame(scores(cca_model_CA2, display = 'bp')) #get the environment vars CC1 and CC2 scores for arrows

cca1_varex<-round(summary(cca_model_CA2)$cont$importance[2,1]*100,2) #Get percentage of variance explained by first axis
cca2_varex<-round(summary(cca_model_CA2)$cont$importance[2,2]*100,2) #Get percentage of variance explained by second axis

# add lifestyle info to sites
df_sites$lifestyle  <- ifelse(grepl("CHE",row.names(df_sites)), "Foragers",
                              ifelse(grepl("NEW00",row.names(df_sites)) | grepl("NEW10",row.names(df_sites)) | grepl("THA",row.names(df_sites)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(df_sites)) | grepl("NEW11",row.names(df_sites)), "Expats", "Recently Settled")))

# fix environmental labels
rownames(df_environ) <- c("alcohol", "tobacco", "smoking", "scarcity", "sex", "location", "yogurt once month", "yogurt twice month", "yogurt twice week", "literacy", "Sisnu", "grain")

scaling_factor_CCA <- 2.5 # for better visualization

final_plot <- ggplot()+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species, color = c("black"), shape = 8, alpha = 0.6, size = 3,
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  # Add species points
  geom_point(data=df_sites, shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  #Add environmental vars arrows
  geom_segment(data=df_environ, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ, size = 3, 
            aes(x=CCA1*scaling_factor_CCA, 
                y=CCA2*scaling_factor_CCA,
                label=rownames(df_environ),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "CCA")

# ggsave(file="CCA.svg", plot=final_plot, width=8, height=7)

# main body CCA plot
df_environ_fig <- df_environ[c("alcohol", "smoking", "location", "Sisnu", "grain"),]

mainbody_plot <- ggplot(data=df_sites)+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add site points
  geom_point(shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  stat_ellipse(aes(x=CCA1, y=CCA2, color = lifestyle),type = "norm") +
  #Add environmental vars arrows
  geom_segment(data=df_environ_fig, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ_fig, size = 3, 
            aes(x=CCA1*scaling_factor_CCA, 
                y=CCA2*scaling_factor_CCA,
                label=rownames(df_environ_fig),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "CCA")
```

# add species info to find taxa that pop out
```{r}
taxa_cutoff <- df_species
taxa_cutoff <- filter(taxa_cutoff, abs(CCA1)>0.5 | abs(CCA2)>0.5)

scaling_factor_CCA <- 2.5

ggplot()+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species, color = c("black"), shape = 8, alpha = 0.6, size = 3,
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  # Add species points
  geom_point(data=df_sites, shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  #Add environmental vars arrows
  geom_segment(data=df_environ, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  # geom_text(data=df_environ, size = 3, 
  #           aes(x=CCA1*scaling_factor_CCA, 
  #               y=CCA2*scaling_factor_CCA,
  #               label=rownames(df_environ),
  #               hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
  #               vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
  #           color="firebrick")+
    geom_text(data=taxa_cutoff, size = 3, 
            aes(x=CCA1, 
                y=CCA2,
                label=rownames(taxa_cutoff),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "CCA")

```

# CCA with CA1
```{r}
set.seed(101)
cca_model_CA1_CA2 <- cca(CCA_otu_tax_t ~ alcohol + tobacco + smoking + scarcity + sex + location + yogurt_freq + literacy + Sisnu + grain + fuel + health_travel + education + ayurvedic + children, data = df_CA_factor)

vif.cca(cca_model_CA1_CA2)

cca_model_CA1_CA2
# 16.04% of species variance explained by the lifestyle factors in the model
# CCA1     CCA2     CCA3     CCA4     CCA5     CCA6     CCA7     CCA8     CCA9    CCA10    CCA11    CCA12 
# 0.029516 0.020865 0.014466 0.010446 0.009610 0.008989 0.008354 0.006594 0.005877 0.005662 0.005158 0.005022 

# test significance of the model
anova.cca(cca_model_CA1_CA2)
#          Df ChiSquare      F Pr(>F)    
# Model    24   0.16038 1.1783  0.013 *
# Residual 38   0.21552   
anova.cca(cca_model_CA1_CA2, by = "terms")
#               Df ChiSquare      F Pr(>F)    
# alcohol        1  0.008906 1.5703  0.035 *  
# tobacco        1  0.006803 1.1996  0.206    
# smoking        1  0.014040 2.4755  0.001 ***
# scarcity       1  0.007700 1.3576  0.070 .  
# sex            1  0.005708 1.0064  0.421    
# location       1  0.012723 2.2434  0.002 ** 
# yogurt_freq    3  0.017730 1.0421  0.378    
# literacy       1  0.004678 0.8249  0.729    
# Sisnu          1  0.011441 2.0173  0.006 ** 
# grain          1  0.008582 1.5133  0.047 *  
# fuel           2  0.011475 1.0116  0.462    
# health_travel  1  0.006568 1.1580  0.211    
# education      3  0.015799 0.9286  0.676    
# ayurvedic      3  0.014646 0.8608  0.808    
# children       3  0.013584 0.7984  0.926   
# Residual      38  0.215515       
anova.cca(cca_model_CA1_CA2, by = "axis")
#          Df ChiSquare      F Pr(>F)   
# CCA1      1  0.029516 5.2042  0.008 **
# CCA2      1  0.020865 3.6790  0.330    

### plot
```

# prep for CCA with CA1 final plot 
```{r}
df_species  <- as.data.frame(summary(cca_model_CA1_CA2)$species[,1:2])# get the species CC1 and CC2 scores
df_sites  <- as.data.frame(summary(cca_model_CA1_CA2)$sites[,1:2])# get the sites CC1 and CC2 scores
df_environ  <- as.data.frame(scores(cca_model_CA1_CA2, display = 'bp')) #get the environment vars CC1 and CC2 scores for arrows

cca1_varex<-round(summary(cca_model_CA1_CA2)$cont$importance[2,1]*100,2) #Get percentage of variance explained by first axis
cca2_varex<-round(summary(cca_model_CA1_CA2)$cont$importance[2,2]*100,2) #Get percentage of variance explained by second axis

# add lifestyle info to sites
df_sites$lifestyle  <- ifelse(grepl("CHE",row.names(df_sites)), "Foragers",
                              ifelse(grepl("NEW00",row.names(df_sites)) | grepl("NEW10",row.names(df_sites)) | grepl("THA",row.names(df_sites)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(df_sites)) | grepl("NEW11",row.names(df_sites)), "Expats", "Recently Settled")))

# fix environmental labels
rownames(df_environ) <- c("alcohol", "tobacco", "smoking", "scarcity", "sex", "location", "yogurt once month", "yogurt twice month", "yogurt twice week", "literacy", "Sisnu", "grain", "fuel elec/gas", "fuel biogas", "health_travel", "education 2-6", "education 7-11", "education 12+", "unknown ayurvedic", "ayurvedic 0-6mo", "ayurvedic 6-12 mo", "1-2 children", "3-5 children", "6+ children")

scaling_factor_CCA <- 2.5 # for better visualization

final_plot <- ggplot()+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species, color = c("black"), shape = 8, alpha = 0.6, size = 3,
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  # Add species points
  geom_point(data=df_sites, shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  #Add environmental vars arrows
  geom_segment(data=df_environ, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ, size = 3, 
            aes(x=CCA1*scaling_factor_CCA, 
                y=CCA2*scaling_factor_CCA,
                label=rownames(df_environ),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "CCA")

# ggsave(file="CCA.svg", plot=final_plot, width=8, height=7)

df_environ_fig <- df_environ[c("alcohol", "smoking", "location", "Sisnu", "grain"),]

mainbody_plot <- ggplot(data=df_sites)+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add site points
  geom_point(shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  stat_ellipse(level = 0.7, aes(x=CCA1, y=CCA2, color = lifestyle),type = "norm") +
  scale_color_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  #Add environmental vars arrows
  geom_segment(data=df_environ_fig, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ_fig, size = 3, 
            aes(x=CCA1*scaling_factor_CCA, 
                y=CCA2*scaling_factor_CCA,
                label=rownames(df_environ_fig),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"))

# ggsave(file="CCA.svg", plot=mainbody_plot, width=5, height=4)
```

# add species info to find taxa that pop out
```{r}
taxa_cutoff <- df_species
taxa_cutoff <- filter(taxa_cutoff, abs(CCA1)>0.5 | abs(CCA2)>0.5)

scaling_factor_CCA <- 3

CCA_taxa <- ggplot()+
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species, color = c("black"), shape = 8, alpha = 0.6, size = 3,
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  # Add species points
  geom_point(data=df_sites, shape = 21, color = c("black"), size = 3, 
            aes(x=CCA1,
                y=CCA2, fill = lifestyle))+
  scale_fill_manual(name=NULL,
                    values=fourcolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats")) +
  #Add environmental vars arrows
  geom_segment(data=df_environ, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor_CCA,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor_CCA), #Ending coordinate in CCA2 
               color="firebrick", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  # Add environmental vars text
  geom_text(data=df_environ, size = 3,
            aes(x=CCA1*scaling_factor_CCA,
                y=CCA2*scaling_factor_CCA,
                label=rownames(df_environ),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow
            color="firebrick")+
    geom_text(data=taxa_cutoff, size = 3, 
            aes(x=CCA1, 
                y=CCA2,
                label=rownames(taxa_cutoff),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="black")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "CCA")

# ggsave(file="CCA_taxa.svg", plot=CCA_taxa, width=8, height=7)
```

# CCA factor vs numeric
```{r}
#### factor
cca_model_CA2 <- cca(CCA_otu_tax_t ~ alcohol + tobacco + smoking + scarcity + sex + location + yogurt_freq + literacy + Sisnu + grain, data = df_CA_final)

vif.cca(cca_model_CA2)

cca_model_CA2
# 23.51% of species variance explained by the lifestyle factors in the model
# CCA1    CCA2    CCA3    CCA4    CCA5    CCA6    CCA7    CCA8    CCA9   CCA10   CCA11   CCA12 
# 0.07409 0.05030 0.03879 0.03302 0.03038 0.02885 0.02573 0.02464 0.02332 0.02157 0.01828 0.01707 

# test significance of the model
anova.cca(cca_model_CA2)
#          Df ChiSquare      F Pr(>F)    
# Model    12   0.38604 1.2807  0.001 ***
# Residual 50   1.25593       
anova.cca(cca_model_CA2, by = "terms")
#             Df ChiSquare      F Pr(>F)    
# alcohol      1   0.03796 1.5113  0.004 ** 
# tobacco      1   0.02806 1.1173  0.170    
# smoking      1   0.04122 1.6409  0.001 ***
# scarcity     1   0.02916 1.1609  0.082 .  
# sex          1   0.02336 0.9301  0.676    
# location     1   0.04782 1.9039  0.001 ***
# yogurt_freq  3   0.08225 1.0915  0.114    
# literacy     1   0.02276 0.9060  0.783    
# Sisnu        1   0.03959 1.5759  0.002 ** 
# grain        1   0.03385 1.3477  0.009 ** 
anova.cca(cca_model_CA2, by = "axis")
#          Df ChiSquare      F Pr(>F)    
# CCA1      1   0.07409 2.9494  0.001 ***
# CCA2      1   0.05030 2.0025  0.019 *  
# CCA3      1   0.03879 1.5441  0.415 

### plot
autoplot(cca_model_CA2)


#### numeric
cca_model_CA2_num <- cca(CCA_otu_tax_t ~ alcohol + tobacco + smoking + scarcity + sex + location + yogurt_freq + literacy + Sisnu + grain, data = df_CA_numeric2)

vif.cca(cca_model_CA2_num)

cca_model_CA2_num
# 22.971% of species variance explained by the lifestyle factors in the model
# CCA1     CCA2     CCA3     CCA4     CCA5     CCA6     CCA7     CCA8     CCA9    CCA10 
# 0.026241 0.016596 0.010756 0.007121 0.006523 0.005259 0.004458 0.003945 0.002999 0.002431 

# test significance of the model
anova.cca(cca_model_CA2_num)
#          Df ChiSquare      F Pr(>F)    
# Model    10  0.086326 1.5502  0.001 ***
# Residual 52  0.289573        
anova.cca(cca_model_CA2_num, by = "terms")
#             Df ChiSquare      F Pr(>F)    
# alcohol      1  0.008906 1.5993  0.020 *  
# tobacco      1  0.006803 1.2217  0.173    
# smoking      1  0.014040 2.5212  0.001 ***
# scarcity     1  0.007700 1.3827  0.077 .  
# sex          1  0.005708 1.0250  0.373    
# location     1  0.012723 2.2848  0.001 ***
# yogurt_freq  1  0.006379 1.1454  0.252    
# literacy     1  0.004824 0.8663  0.641    
# Sisnu        1  0.011446 2.0554  0.003 ** 
# grain        1  0.007798 1.4002  0.068 .  
# Residual    52  0.289573 
anova.cca(cca_model_CA2_num, by = "axis")
#          Df ChiSquare      F Pr(>F)    
# CCA1      1  0.026241 4.7122  0.001 ***
# CCA2      1  0.016596 2.9802  0.012 *  
# CCA3      1  0.010756 1.9314  0.458 

### plot
p1 <- autoplot(cca_model_CA2_num, title = "Autoplot Numeric")
p2 <- autoplot(cca_model_CA2, title = "Autoplot Categorical")

### categorical
# Get CCA scores
df_species  <- as.data.frame(summary(cca_model_CA2)$species[,1:2])# get the species CC1 and CC2 scores
df_environ  <- as.data.frame(scores(cca_model_CA2, display = 'bp')) #get the environment vars CC1 and CC2 scores for arrows

cca1_varex<-round(summary(cca_model_CA2)$cont$importance[2,1]*100,2) #Get percentage of variance explained by first axis
cca2_varex<-round(summary(cca_model_CA2)$cont$importance[2,2]*100,2) #Get percentage of variance explained by second axis

scaling_factor <- 2 # for better visualization

p3 <- ggplot(df_species, 
       aes(x=CCA1, y=CCA2)) + 
  #Draw lines on x = 0 and y = 0
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species, color = c("forestgreen"),
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  #Add environmental vars arrows
  geom_segment(data=df_environ, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor), #Ending coordinate in CCA2 
               color="firebrick1", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ, 
            aes(x=CCA1*scaling_factor, 
                y=CCA2*scaling_factor,
                label=rownames(df_environ),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick1")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex," %)"),
       y=paste0("CCA2 (",cca2_varex," %)"),
       title = "ggplot2 Categorical")

#### numeric
# Get CCA scores
df_species_num  <- as.data.frame(summary(cca_model_CA2_num)$species[,1:2])# get the species CC1 and CC2 scores
df_environ_num  <- as.data.frame(scores(cca_model_CA2_num, display = 'bp')) #get the environment vars CC1 and CC2 scores for arrows

cca1_varex_num<-round(summary(cca_model_CA2_num)$cont$importance[2,1]*100,2) #Get percentage of variance explained by first axis
cca2_varex_num<-round(summary(cca_model_CA2_num)$cont$importance[2,2]*100,2) #Get percentage of variance explained by second axis

scaling_factor <- 2 # for better visualization

p4 <- ggplot(df_species_num, 
       aes(x=CCA1, y=CCA2)) + 
  #Draw lines on x = 0 and y = 0
  geom_hline(yintercept=0, 
             linetype="dashed") +
  geom_vline(xintercept=0, 
             linetype="dashed") +
  coord_fixed()+
  # Add species points
  geom_point(data=df_species_num, color = c("forestgreen"),
            aes(x=CCA1,#Score in CCA1 to add species text
                y=CCA2,#Score in CCA2 to add species text
                # label=rownames(df_species),
                # hjust=0.5*(1-sign(CCA1)),#Set the text horizontal alignment according to its position in the CCA plot
                # vjust=0.5*(1-sign(CCA2))),#Set the text vertical alignment according to its position in the CCA plot
            ))+
  #Add environmental vars arrows
  geom_segment(data=df_environ_num, 
               aes(x=0, #Starting coordinate in CCA1 = 0 
                   xend=CCA1*scaling_factor,#Ending coordinate in CCA1  
                   y=0, #Start in CCA2 = 0
                   yend=CCA2*scaling_factor), #Ending coordinate in CCA2 
               color="firebrick1", #set color
               arrow=arrow(length=unit(0.01,"npc"))#Set the size of the lines that form the tip of the arrow
               )+
  #Add environmental vars text
  geom_text(data=df_environ_num, 
            aes(x=CCA1*scaling_factor, 
                y=CCA2*scaling_factor,
                label=rownames(df_environ_num),
                hjust=0.5*(1-sign(CCA1)),#Add the text of each environmental var at the end of the arrow
                vjust=0.5*(1-sign(CCA2))),#Add the text of each environmental var at the end of the arrow 
            color="firebrick1")+
  #Set bw theme
  theme_bw()+
  #Set x and y axis titles
  labs(x=paste0("CCA1 (",cca1_varex_num," %)"),
       y=paste0("CCA2 (",cca2_varex_num," %)"),
       title = "ggplot2 Numeric")


grid.arrange(p1, p2, p3, p4, ncol = 2)

```

